wallet: PSBT channel funding

guggero opened the pull request:
Fixes #3085 and #3936.

This PR introduces channel funding through PSBTs! With this, it is possible to open one or more channels from `lnd` without it needing to have a wallet balance.

Roasbeef commented on go.mod:
```diff
@@ -45,7 +46,7 @@ require (
 	github.com/tv42/zbase32 v0.0.0-20160707012821-501572607d02
 	github.com/urfave/cli v1.18.0
-	golang.org/x/crypto v0.0.0-20200109152110-61a87790db17
+	golang.org/x/crypto v0.0.0-20200115085410-6d4e4cb37c7d
```
Any way we can suppress this?

guggero replied:
Good catch!
It seems this is added automatically on running `go get ./...` because the new version of `btcsuite/btcutil` that contains the PSBT stuff references it. I added an explicit overwrite to prevent using a version that we haven't verified.

Roasbeef commented on lnrpc/rpc.proto:
```diff
@@ -1703,6 +1703,15 @@ message PendingUpdate {
 }
 
+message ReadyForPsbtFunding {
+    string funding_address = 1;
```
How about making this a one-of that also includes a raw PSBT? Just to support all the flows we have in mind.

guggero replied:
I changed the response to always contain the address, amount and PSBT so the user has the choice of what to use. A `oneof` doesn't really make sense IMO as it's the same information, just encoded differently twice.

Roasbeef commented on lnrpc/rpc.proto:
```diff
@@ -1774,8 +1783,23 @@ message OpenChannelRequest {
+        */
         ChannelOpenUpdate chan_open = 3;
+
+        /**
```
Relying on the streaming response may make this harder to use over REST. Have you considered instead making the workflow primarily drive a state machine via `FundingStateStep`? Either way though, it seems like we need a new response here.

guggero replied:
I think this can be done by adding a new RPC `OpenChannelPSBTSync` that works very similar to `OpenChannelSync` but already returns once the `OpenStatusUpdate_PsbtFund` message arrives.
The rest can then be done through the transition messages. I'll add a status type to the funding intent to turn it into a "real" state machine and add a new message like `FundingTransitionMsg_Status` to query it.
I should have everything ready by tomorrow.

Roasbeef replied:
Haven't reviewed the latest version yet, but why couldn't it just use `OpenChannelSync` as is? In any case, REST is secondary, and this is a newer experimental API, so I don't think it's a deal breaker if the initial version doesn't support REST.

Roasbeef commented on lnrpc/rpc.proto:
```diff
@@ -1830,9 +1854,35 @@ message ChanPointShim {
         ChanPointShim chan_point_shim = 1;
+        
+        /**
+        A channel shim that uses a PSBT to fund and sign the channel funding
```
So the idea is that this will be specified when _initiating_ the channel opening right?

guggero replied:
Yes. I was confused about the use of the `shim_register` in `FundingStateStep`. See comment further below.

Roasbeef commented on lnrpc/rpc.proto:
```diff
@@ -1830,9 +1854,35 @@ message ChanPointShim {
+    should only be set if the responder is already expecting a specific pending
+    channel ID.
+    */
+    bytes pending_chan_id = 2;
```
Not sure we need this, but I guess it can't hurt. 

One other thing we could add here, is a serialized PSBT to serve as the base for the transaction itself. This could be used to progressively create a set of combined channel openings across several nodes. This isn't blocking in my mind though.

guggero replied:
It's used to identify which shim we are referring to. To validate the PSBT we need to know the amount and multisig script that we expect to be in there. We can only get these through the shim which is stored in the reservation.

Roasbeef commented on lnrpc/rpc.proto:
```diff
@@ -1703,6 +1703,15 @@ message PendingUpdate {
+    int64 funding_amount = 2;
+}
+
+message ReadyForPsbtSigning {
```
Is this used anywhere?

guggero replied:
Yes, this is returned in the stream answer to the `OpenChannel` call.

Roasbeef commented on lnwallet/chanfunding/psbt_assembler.go:
```diff
@@ -0,0 +1,135 @@
+	Psbt *psbt.Packet
+}
+
+func (i *PsbtIntent) CompileFundingTx() (*wire.MsgTx, error) {
```
Missing `godoc` comments here.

Roasbeef commented on lnwallet/chanfunding/psbt_assembler.go:
```diff
@@ -0,0 +1,135 @@
+func (i *PsbtIntent) BindKeys(localKey *keychain.KeyDescriptor,
+	remoteKey *btcec.PublicKey) {
+
+	i.localKey = localKey
```
Is this needed given that we wrap the shim intent?

guggero replied:
Yes, the `BindKeys` only also exists on the `FullIntent`, not the `ShimIntent`.

Roasbeef commented on lnwallet/chanfunding/psbt_assembler.go:
```diff
@@ -0,0 +1,135 @@
+
+// NewPsbtAssembler creates a new CannedAssembler from the material required
+// to construct a funding output and channel point.
+func NewPsbtAssembler(fundingAmt btcutil.Amount) *PsbtAssembler {
```
Do we also need a way to pass in a partial PSBT here?

Roasbeef commented on fundingmanager.go:
```diff
@@ -1547,6 +1548,37 @@ func (f *fundingManager) handleFundingAccept(fmsg *fundingAcceptMsg) {
 	}
 	err = resCtx.reservation.ProcessContribution(remoteContribution)
+
+	// The wallet has detected that a PSBT funding process was requested by
```
Still getting through this, but we could possibly attempt to encapsulate more of this in the reservation struct itself.

Roasbeef replied:
Or I guess the idea is that this method will end up being called _twice_ for PSBT based funding?

Roasbeef replied:
One alternative would be to create a new goroutine for this specific purpose which will then resume things on its own, independent of the main funding manager control loop.

guggero replied:
I like the idea with the goroutine. I went down that road and implemented it. I didn't have time to fully test it yet but it should work like that. I'll continue on it tomorrow.

Roasbeef commented on lnwallet/wallet.go:
```diff
@@ -907,6 +1004,77 @@ func (l *LightningWallet) handleContributionMsg(req *addContributionMsg) {
 			"generated: %v", chanPoint, spew.Sdump(fundingTx))
 	}
 
+	// This is a normal flow where no PSBT step is involved, so we can
```
The comment says no PSBT step is involved, but the referenced method has PBST related logic.

guggero replied:
Fixed the comment.

Roasbeef commented on rpcserver.go:
```diff
@@ -1684,6 +1701,31 @@ func (r *rpcServer) OpenChannel(in *lnrpc.OpenChannelRequest,
+			// The funding intent must be present.
+			psbtShim := in.FundingShim.GetPsbtShim()
+			copy(req.pendingChanID[:], psbtShim.PendingChanId)
+			intent, err := r.server.cc.wallet.GetFundingIntent(
```
So we require the user to already have registered the intent? Usually we require registration only when we're the _responder_ to a special funding flow. Instead, I would think this step would just set the channel funder to the PSBT _assembler_ itself.

guggero replied:
Ah, I did indeed ask myself why the registration of the shim was needed prior to opening a channel when we send it there again. I didn't think of the case were we _aren't_ the initiator and still need a way to inform the funding manager about the incoming channel. It makes much more sense now and I removed the requirement for the shim to be registered before calling `OpenChannel`.

Roasbeef commented on rpcserver.go:
```diff
@@ -6051,6 +6093,40 @@ func (r *rpcServer) FundingStateStep(ctx context.Context,
 			return nil, err
 		}
 
+	// If this is a message to register a new shim that should be funded
```
Same as above, wouldn't expect the user to themselves need to register for a specific PSBT shim up front. But perhaps you have some new workflows in mind?

Roasbeef commented on cmd/lncli/psbt.go:
```diff
@@ -0,0 +1,283 @@
+// protocol involves several steps between the RPC server and the CLI client:
+//
+// RPC server                           CLI client
+//     |<----register funding shim----------|
```
So here's my core question again: why does the user need to register the funding shim _before_ they open the channel? Compared to just passing it in with the original funding request (RPC level).

Roasbeef replied:
Other than this, the external RPC flow looks sound IMO.

guggero replied:
Fixed it by removing the initial register call.

Roasbeef commented on cmd/lncli/psbt.go:
```diff
@@ -0,0 +1,283 @@
+	req.FundingShim = &lnrpc.FundingShim{
+		Shim: &lnrpc.FundingShim_PsbtShim{
+			PsbtShim: &lnrpc.PsbtShim{
+				PendingChanId: pendingChanID[:],
```
Don't think we need to make them specify the pending channel ID here. Instead things can go directly to using open to kick things off.

Roasbeef commented on lnrpc/rpc.proto:
```diff
@@ -1900,9 +1937,41 @@ message ChanPointShim {
+    The size of the output to be created as the channel point for this channel
+    funding.
+    */
+    int64 amt = 1;
```
Is this needed given that the greater request already has the amount specified?

guggero replied:
No, removed.

Roasbeef commented on lnrpc/rpc.proto:
```diff
@@ -1911,17 +1980,56 @@ message FundingShimCancel {
+    capacity amount to the PK script returned in the open channel message in a
+    previous step.
+    */
+    bytes signed_psbt = 1;
```
To make the API harder to use in an invalid manner, what if we just accepted the signatures/witnesses here, and then execute the finalization steps ourselves internally? Otherwise, it feels like the user can just skip straight to this step (assuming it still does verification).

Roasbeef commented on lnwallet/chanfunding/psbt_assembler.go:
```diff
@@ -0,0 +1,372 @@
+
+// Verify makes sure the PSBT that is given to the intent has an output that
+// sends to the channel funding multisig address with the correct amount. A
+// simple check that there is a large enough input present is also performed.
```
> there is a large enough input present is also performed

I think this instead should be: "that at least a single input has been specified is performed". Assuming we don't have access to the UTXO state in this context, we can't check the input UTXO size.

guggero replied:
Right, thanks. I wrote the comment before I realized we don't have the UTXO information here.

Roasbeef commented on lnwallet/chanfunding/psbt_assembler.go:
```diff
@@ -0,0 +1,372 @@
+		return fmt.Errorf("funding output cannot be created: %v", err)
+	}
+	outputFound := false
+	for _, out := range packet.UnsignedTx.TxOut {
```
We should also check for the proper output value as well.

guggero replied:
That's why I'm using `DeepEqual` so both the output value and PK script are checked. Or what am I missing?

Roasbeef commented on lnwallet/chanfunding/psbt_assembler.go:
```diff
@@ -0,0 +1,372 @@
+
+// Finalize makes sure the final PSBT that is given to the intent is fully valid
+// and signed but still contains the same UTXOs and outputs as the pending
+// transaction we previously verified. If everything checks out, the funding
```
Ah ok, so we will in fact verify again, you can safely ignore my prior comment then.

Roasbeef commented on lnwallet/chanfunding/psbt_assembler.go:
```diff
@@ -0,0 +1,372 @@
+		return fmt.Errorf("number of outputs differ from verified PSBT")
+	}
+	for idx, out := range packet.UnsignedTx.TxOut {
+		if !reflect.DeepEqual(out, verifiedOuts[idx]) {
```
Similar thing here about output value (can likely be coupled into a helper function). Also, if we want to go even further here, we could do a full equality check (removed witness/sigScript) between this an the prior one we extracted.

Roasbeef commented on lnwallet/chanfunding/psbt_assembler.go:
```diff
@@ -0,0 +1,372 @@
+	if err != nil {
+		return fmt.Errorf("error finalizing PSBT: %v", err)
+	}
+	if !packet.IsComplete() {
```
This can be omitted, as the first thing `Extract` does is ensure the packet is complete: https://github.com/btcsuite/btcutil/blob/master/psbt/extractor.go#L24

Roasbeef commented on lnwallet/chanfunding/psbt_assembler.go:
```diff
@@ -0,0 +1,372 @@
+//
+// NOTE: Part of the chanfunding.Intent interface.
+func (i *PsbtIntent) Cancel() {
+	i.PsbtReady <- fmt.Errorf("user canceled funding flow")
```
Should make sure this channel is _always_ buffered.

Roasbeef commented on lnwallet/chanfunding/psbt_assembler.go:
```diff
@@ -0,0 +1,372 @@
+		},
+		State:     PsbtShimRegistered,
+		BasePsbt:  p.basePsbt,
+		PsbtReady: make(chan error),
```
Should buffer with a size of 1.

Roasbeef commented on fundingmanager.go:
```diff
@@ -1559,6 +1589,90 @@ func (f *fundingManager) handleFundingAccept(fmsg *fundingAcceptMsg) {
 	fndgLog.Debugf("Remote party accepted commitment constraints: %v",
 		spew.Sdump(remoteContribution.ChannelConfig.ChannelConstraints))
 
+	// If the user requested funding through a PSBT, we cannot directly
```
👍

Roasbeef commented on rpcserver.go:
```diff
@@ -1551,6 +1552,39 @@ func newFundingShimAssembler(chanPointShim *lnrpc.ChanPointShim, initiator bool,
+		packet *psbt.Packet
+		err    error
+	)
+	// The base PSBT is optional. But if it's set, it has to be a valid,
```
Style nit: missing new line above.

Roasbeef commented on rpcserver.go:
```diff
@@ -1686,6 +1721,49 @@ func (r *rpcServer) OpenChannel(in *lnrpc.OpenChannelRequest,
+			// We'll also need to inform the wallet about the
+			// funding intent, so ask the assembler to provision it
+			// for us.
+			psbtIntent, err := psbtAssembler.ProvisionChannel(
```
Does this need to be provisioned _outside_ the wallet? Or is it that the later steps depend on the wallet already having stored the intent in the internal map for retrieval purposes?

guggero replied:
Ah, no, it's not needed. This was left in from my misunderstanding of how `GetShimRegister` was used.

Roasbeef commented on rpcserver.go:
```diff
@@ -6179,18 +6257,92 @@ func (r *rpcServer) FundingStateStep(ctx context.Context,
-		err := r.server.cc.wallet.CancelFundingIntent(pendingChanID)
+		rpcsLog.Debugf("Verifying PSBT for pending_id=%x",
+			pendingChanID)
+		intent, err := r.server.cc.wallet.GetFundingIntent(
```
In self-response to my comment above, perhaps this can be pushed down further into the wallet? Which would then mean we don't really need the `GetFundingIntent` method as the wallet is able to look it up itself by reservation (with the added methods for the reservation being added). Don't feel strongly about either path though in the end.

guggero replied:
No, you're right. It absolutely makes sense to move this, makes the code much more clean.

Roasbeef commented on lntest/itest/psbt.go:
```diff
@@ -0,0 +1,308 @@
+	"github.com/lightningnetwork/lnd/lntest"
+)
+
+func testPsbtChanFunding(net *lntest.NetworkHarness, t *harnessTest) {
```
Great itest 👍

guggero commented:
I further cleaned up the logic of the PR, thanks for the suggestions, @Roasbeef.
I also added two new commits that fix problems that I found while doing some manual tests.

wpaulino commented on fundingmanager.go:
```diff
@@ -589,6 +589,11 @@ func (f *fundingManager) Stop() error {
 func (f *fundingManager) stop() error {
 	fndgLog.Infof("Funding manager shutting down")
 
+	// Before whe stop, make sure we give the remote peer a chance to clean
```
```suggestion
	// Before we stop, make sure we give the remote peer a chance to clean
```

wpaulino commented on fundingmanager.go:
```diff
@@ -3307,6 +3312,58 @@ func (f *fundingManager) pruneZombieReservations() {
+// for reservations where we artificially stalled/paused the progress because
+// the user needs to perform manual funding steps (like signing a PSBT for
+// example).
+func (f *fundingManager) failPausedReservations() {
```
This looks fine to me, but I'm wondering if it's necessary? I would imagine that the responding node would just forget of the incomplete reservation eventually. Did you run into any issues while testing that required this?

guggero replied:
Yes, I noticed that if I shut down `lnd` while a PSBT funding was in progress (to test that no lingering channel ready somewhere block shutdown) I noticed that you then have to wait 10 minutes until you can try opening a channel to the same node as before again, otherwise you would run into "maximum number of pending channels exceeded" error. But it's quite unlikely this happens, I agree. I'm fine with removing it again.

guggero replied:
Ah, you are right. I tested it again and there was no wait. Must have been something else I didn't clean up in my first tests.
I looked into this again and tried canceling all reservations. But at the moment of the shutdown of the funding manager, the wallet has already stopped its message consuming loop and we get a deadlock.
And since we know that the reservation and coins locks are in-memory only on our side it's probably OK to do nothing on shutdown, as it was before.

wpaulino commented on docs/psbt.md:
```diff
@@ -0,0 +1,205 @@
+example simple. Of course any other PSBT compatible wallet could be used and the
+process would likely be spread out over multiple signing steps. The goal of this
+example is not to cover each and every possible edge case but to help users of
+`lnd` to understand what inputs the `lncli` utility expects.
```
```suggestion
`lnd` understand what inputs the `lncli` utility expects.
```

wpaulino commented on docs/psbt.md:
```diff
@@ -0,0 +1,205 @@
+a PSBT. That means, `lnd` can have a wallet balance of `0` and is still able to
+open a channel. Let's start.
+
+### 1. use the new `--psbt` flag in `lncli openchannel`
```
```suggestion
### 1. Use the new `--psbt` flag in `lncli openchannel`
```

wpaulino commented on docs/psbt.md:
```diff
@@ -0,0 +1,205 @@
+change output for the rest. Everything is there but the signatures/witness data,
+which is exactly what we need.
+
+### 3. verify and sign the PSBT
```
```suggestion
### 3. Verify and sign the PSBT
```

wpaulino commented on cmd/lncli/psbt.go:
```diff
@@ -0,0 +1,311 @@
+//     |                                    |
+func openChannelPSBT(ctx *cli.Context, client lnrpc.LightningClient,
+	req *lnrpc.OpenChannelRequest, basePsbt string, block bool) error {
+
```
We should reject any flags that are incompatible with the PSBT flow, e.g., fees.

wpaulino commented on fundingmanager.go:
```diff
@@ -70,6 +70,13 @@ const (
+	// there are manual steps and possibly multiple devices involved, the
+	// default timeout of 10 minutes is likely not enough for certain use
+	// cases. We assume that one hour is probably more realistic.
+	ReservationTimeoutMultiplierPSBT = 6
```
Have you tested against any other implementations to see how they would react to the longer execution time throughout the PSBT flow?

guggero replied:
No, I haven't tried that yet, need to add other implementations to my setup.
But I tried against another `lnd` and that node canceled my flow after 10 minutes :(
Apparently it's not enough if the initiating node waits an hour, the remote node also needs to cooperate... Should we start increasing the default time of 10 minutes with `0.10.0` to relax this at least for `lnd` nodes? Or should we look into "keep alive" mechanisms?

Roasbeef replied:
Yeah correct that _both_ nodes need to have this higher value. We can set it higher, but we'll still need to handle the timeout case, so the user should get a proper error that things went wrong and they need to try again. As the remote party will likely se a distinct funding key, they'll need to repeat the whole process again.

guggero replied:
It doesn't make sense to increase the timeout only for PSBT flows. The other node will never really know we're using that funding flow. If we're going to increase anything, it should be the default 10 minutes to something like 30 minutes perhaps.
I removed this multiplier again.

guggero commented on rpcserver.go:
```diff
@@ -6207,15 +6260,51 @@ func (r *rpcServer) FundingStateStep(ctx context.Context,
+	// even though our RPC message structure allows for it. Inform the user
+	// by returning a proper error instead of just doing nothing.
+	case in.GetShimRegister() != nil &&
+		in.GetShimRegister().GetPsbtShim() != nil:
```
No, I don't see a use case where this would be needed. This is added to catch incorrect API usage and inform the user they're doing something wrong instead of just returning an empty message.

guggero commented on cmd/lncli/psbt.go:
```diff
@@ -0,0 +1,311 @@
+//     |  |-------channel open------------->|
+//     |                                    |
+func openChannelPSBT(ctx *cli.Context, client lnrpc.LightningClient,
+	req *lnrpc.OpenChannelRequest, basePsbt string, block bool) error {
```
I wanted to avoid the `psbt.go` file containing hard coded strings of flag names as they would likely be forgotten there if we ever changed them in the main file.

guggero commented on lnwallet/chanfunding/psbt_assembler.go:
```diff
@@ -0,0 +1,399 @@
+func (p *PsbtAssembler) ProvisionChannel(req *Request) (Intent, error) {
+	// We'll exit out if this field is set as the funding transaction has
+	// already been assembled, so we don't influence coin selection..
+	if req.SubtractFees {
```
We can't check the fee rate here as we're getting fed in a default rate. But I added a check at the RPC level and in the command line.

guggero commented on lnwallet/chanfunding/psbt_assembler.go:
```diff
@@ -0,0 +1,399 @@
+	// deeper into input validation for the moment as that involves many
+	// intricacies. We see it as the input adder's job to make sure the
+	// amount is sufficient and that a change output is added.
+	if len(packet.UnsignedTx.TxIn) == 0 {
```
I looked into it a bit further and it's not as complicated as I first thought. Added the check.

guggero commented on cmd/lncli/psbt.go:
```diff
@@ -0,0 +1,311 @@
+	go func() {
+		<-signal.ShutdownChannel()
+		close(quit)
+		maybeCancelShim()
```
In most cases, yes. But if we are waiting for a message to arrive, just closing the quit channel is not enough, we need to cancel the context. I added some comments and wrapped it in a `Once`.

guggero commented:
Thanks for all the good feedback, @wpaulino! I reworked a bunch of code and addressed all your comments.

One question that was brought up in the comments is what happens if we wait longer than 10 minutes, either with the other peer being another implementation or `lnd` of any previous version.
I confirmed that if we open a channel with a peer of `lnd` version `v0.9.2` and below, we will get the funding reservation canceled after 10 minutes of being idle. Cause is the `pruneZombieReservations` cleanup method.
I'm not really sure what we can do here.
I see three possible options:
1. Tell the user they have 10 minutes max and that they need to hurry. Not the best UX.
2. Send some kind of message to the other peer that triggers an update of the `resCtx.lastUpdated` timestamp. No idea what message would work that doesn't modify any other state.
3. Increase the default zombie pruning interval from 10 minutes to 60 minutes or something similar. But that would only solve the problem when opening channels to peers with version `v0.10.0` or later.

Any other ideas what we could do here?

Roasbeef commented on lnwallet/chanfunding/psbt_assembler.go:
```diff
@@ -0,0 +1,490 @@
+		case in.NonWitnessUtxo != nil:
+			// Non-witness UTXOs reference to the whole transaction
+			// the UTXO resides in.
+			utxOuts := in.NonWitnessUtxo.TxOut
```
Nice, I totally forgot that the inputs in a PSBT include the value as well, super convenient for verification checks like this.

Roasbeef commented on rpcserver.go:
```diff
@@ -6207,15 +6274,51 @@ func (r *rpcServer) FundingStateStep(ctx context.Context,
+	case in.GetPsbtVerify() != nil:
+		rpcsLog.Debugf("Verifying PSBT for pending_id=%x",
+			in.GetPsbtVerify().PendingChanId)
+		copy(pendingChanID[:], in.GetPsbtVerify().PendingChanId)
```
Style nit: could use some spacing here between these lines, how here and below.

Roasbeef commented on fundingmanager.go:
```diff
@@ -1559,6 +1607,85 @@ func (f *fundingManager) handleFundingAccept(fmsg *fundingAcceptMsg) {
+	// funding transaction. If the channel is closed without any error being
+	// sent, we know everything's going as expected.
+	select {
+	case err := <-intent.PsbtReady:
```
If we add another case here where we listen on a channel that's closed when this reservation is also cancelled, then this should be enough to handle the case where the remote party fails the funding flow if we're taking too long. So in this case, at least the user ill get a nice error that hey need to try again.

guggero commented on lnrpc/rpc.proto:
```diff
@@ -1900,9 +1937,35 @@ message ChanPointShim {
+    protocol to initiate the funding request. This is an optional field, and
+    should only be set if the responder is already expecting a specific pending
+    channel ID.
+    */
```
Thanks, good catch!

guggero commented on lnwallet/chanfunding/psbt_assembler.go:
```diff
@@ -0,0 +1,504 @@
+	if err != nil {
+		return fmt.Errorf("outputs differ from verified PSBT: %v", err)
+	}
+	err = verifyInputPrevOutpointsEqual(
```
Good idea. I'll add them to the library as a follow-up PR.

guggero commented on fundingmanager.go:
```diff
@@ -3144,6 +3275,17 @@ func (f *fundingManager) handleErrorMsg(fmsg *fundingErrorMsg) {
+	// If this was a PSBT funding flow, the remote likely timed out because
+	// we waited too long. Return a nice error message to the user in that
+	// case so the user knows what's the problem.
+	if resCtx.reservation.IsPsbt() {
```
You're right, the original error shouldn't be suppressed. I just wrapped the error with the "remote canceled, possibly timed out" message instead of replacing it.

guggero commented:
About the logs: Yes, that is too much. There was still an error in there that if the remote peer canceled the reservation, we also tried to cancel it back, which resulted in additional errors.
I fixed the error and demoted some messages to debug.

wpaulino commented on docs/psbt.md:
```diff
@@ -0,0 +1,216 @@
+
+## Opening a channel by using a PSBT
+
+This is a step-by-step instruction on how to open a channel with `lnd` by using
```
```suggestion
This is a step-by-step guide on how to open a channel with `lnd` by using
```

wpaulino commented on docs/psbt.md:
```diff
@@ -0,0 +1,216 @@
+a PSBT. That means, `lnd` can have a wallet balance of `0` and is still able to
+open a channel. We'll jump into an example right away.
+
+But first we need to mention a small caveat of this new funding flow: _Time
```
```suggestion
This new funding flow has a small caveat: _time
```

wpaulino commented on docs/psbt.md:
```diff
@@ -0,0 +1,216 @@
+
+But first we need to mention a small caveat of this new funding flow: _Time
+matters_.  
+When opening a channel using the PSBT flow, we start the negotiation
```
Nit: either introduce a newline before or merge this line with the one above.

wpaulino commented on docs/psbt.md:
````diff
@@ -0,0 +1,216 @@
+Base64 encoded PSBT:
+```
+
+The command line now waits until a PSBT is pasted in there. We'll create one
````
```suggestion
The command line now waits until a PSBT is entered. We'll create one
```

wpaulino commented on cmd/lncli/cmd_open_channel.go:
```diff
@@ -0,0 +1,277 @@
+package main
```
This commit (7f96272b84490c7b4507307d6e64a24b1f0b60f6) is showing out of order, ideally it comes before the PSBT one.

guggero replied:
Yes, it's before. Was just showing incorrectly on GitHub because of the time based sort. Fixed the commit timestamps.

guggero commented:
Nits fixed, commit dates adjusted and rebased.
